syntax = "proto3";
package WasteCollection;

// WasteCollectionService
service WasteCollectionService{

    // GetRoute Procedure: Determines a route for a waste vehicle to take.
    rpc GetRoute(VehicleIdentifier) returns (CollectionRoute);

    // DispatchAnotherVehicle: When a waste vehicle reaches max capacity, request another vehicle to finish the route.
    rpc DispatchAnotherVehicle(VehicleCurrentContext) returns (DispatchInfo);

}

// VehicleIdentifier consists of the vehicle's ID and waste collection type.
message VehicleIdentifier{
    int32 vehicle_id = 1;
    string vehicle_type = 2;
}

// CollectionRoute is a list of directed nodes representing a collection route.
message CollectionRoute{
    repeated int32 nodes = 1;
}

// VehicleCurrentContext consists of the vehicle's ID, type, and remaining route.
message VehicleCurrentContext{

    int32 vehicle_id = 1;
    string vehicle_type = 2;
    repeated int32 remaining_nodes = 3;
}

// DispatchInfo consists of the new vehicle's ID, type, and starting node.
message DispatchInfo{

    int32 vehicle_id = 1;
    string vehicle_type = 2;
    int32 startNode = 3;

}